src/glue/jobs/json_to_parquet.py

```diff
@@ -182,11 +183,86 @@ def drop_deleted_healthkit_data(
     )
     return table_with_deleted_samples_removed
 
+def archive_existing_datasets(
+        glue_context: GlueContext,
+        bucket: str,
+        key_prefix: str,
+        workflow_run_id: str,
+        delete_upon_completion: bool
+    ) -> None:
+    """
+    Archives existing datasets in S3 by copying them to a timestamped subfolder
+    within an "archive" folder. The format of the timestamped subfolder is:
+
+        {year}_{month}_{day}_{workflow_run_id}
+
+    If the dataset does not exist at `key_prefix` or the dataset is empty,
+    no action will be taken.
+
+    Args:
+        glue_context (GlueContext): The Glue context
+        bucket (str): The name of the S3 bucket containing the dataset to archive.
+        key_prefix (str): The S3 key prefix of the objects to be archived.
+            Ensure that this ends with a trailing "/" to avoid matching
+            child datasets.
+        workflow_run_id (str): The Glue workflow run ID.
+        delete_upon_completion (bool): Whether to delete the dataset after archiving.
+
+    Returns:
+        None
+
+    Raises:
+        N/A
+    """
+    logger = glue_context.get_logger()
+    s3_client = boto3.client("s3")
+    response = s3_client.list_objects_v2(
+            Bucket=bucket,
+            Prefix=key_prefix
+    )
+    if "Contents" not in response:
+        return
+    formatted_datetime = ( # e.g., 2023_06_28
+            str(datetime.now())
+            .split(" ")[0]
+            .replace('-', '_')
+    )
+    this_archive_collection = "_".join([formatted_datetime, workflow_run_id])
+    source_objects = [
+            obj for obj in response["Contents"] if obj["Size"] > 0]
+    objects_to_delete = []
+    for source_object in source_objects:
+        this_object = {
+                "Bucket": bucket,
+                "Key": source_object["Key"]
+        }
+        destination_key = os.path.join(
+                os.path.dirname(key_prefix[:-1]), # .../parquet/
+                "archive",
+                this_archive_collection, # {year}_{month}_{day}_{workflow_run_id}
+                os.path.basename(key_prefix[:-1]), # {dataset}
+                os.path.relpath(this_object["Key"], key_prefix) # some/path/object.parquet
+        )
+        logger.info(f"Copying {this_object['Key']} to {destination_key}")
+        copy_response = s3_client.copy_object(
+                CopySource=this_object,
+                Bucket=bucket,
+                Key=destination_key
+        )
+        this_object.pop("Bucket")
+        objects_to_delete.append(this_object)
+    if delete_upon_completion:
+        s3_client.delete_objects(
+                Bucket=bucket,
+                Delete={"Objects": objects_to_delete}
```

Comment:
Will this line of code fail if `object_to_delete` is empty?

Reply:
`object_to_delete` might be empty if the data type key prefix exists but there are no objects in it? It does throw an error if `object_to_delete` is empty so I added the condition that it must not be empty.